src/main/scala/io/iohk/ethereum/blockchain/sync/SyncStateSchedulerActor.scala

```diff
@@ -97,73 +167,298 @@ class SyncStateSchedulerActor(downloader: ActorRef, sync: SyncStateScheduler, sy
+    log.debug("Starting request sequence")
+    sync.persistBatch(currentState, targetBlock)
+    restartRequester ! WaitingForNewTargetBlock
+    context.become(idle(currentStats.addSaved(currentState.memBatch.size)))
+  }
+  import akka.pattern.pipe
+
+  // scalastyle:off
   def syncing(
       currentState: SchedulerState,
+      currentDownloaderState: DownloaderState,
       currentStats: ProcessingStatistics,
       targetBlock: BigInt,
-      syncInitiator: ActorRef
-  ): Receive = {
-    case MissingNodes(nodes, downloaderCap) =>
-      log.debug(s"Received {} new nodes to process", nodes.size)
-      // Current SyncStateDownloaderActor makes sure that there is no not requested or duplicated values in its response.
-      // so we can ignore those errors.
-      //TODO [ETCM-275] process responses asynchronously to avoid steep rise of pending requests after pivot block update
-      sync.processResponses(currentState, nodes) match {
-        case Left(value) =>
-          log.error(s"Critical error while state syncing ${value}, stopping state sync")
-          // TODO we should probably start sync again from new target block, as current trie is malformed or declare
-          // fast sync as failure and start normal sync from scratch
-          context.stop(self)
-        case Right((newState, statistics)) =>
-          if (newState.numberOfPendingRequests == 0) {
-            reportStats(syncInitiator, currentStats, currentState)
-            finalizeSync(newState, targetBlock, syncInitiator)
-          } else {
-            log.debug(
-              s" There are {} pending state requests," +
-                s"Missing queue size is {} elements",
-              newState.numberOfPendingRequests,
-              newState.queue.size()
+      syncInitiator: ActorRef,
+      nodesToProcess: Queue[RequestResult],
+      processing: Boolean,
+      restartRequested: Option[ActorRef]
+  ): Receive = handleCommonMessages orElse handleRequestResults orElse {
+    case Sync if currentState.numberOfPendingRequests > 0 && restartRequested.isEmpty =>
+      val freePeers = getFreePeers(currentDownloaderState)
+      nodesToProcess.dequeueOption match {
+        case Some((nodes, restOfRequests)) if freePeers.nonEmpty =>
+          log.debug(
+            "Got {} peer responses to Process, and there are {} idle peers available",
+            restOfRequests.size,
+            freePeers.size
+          )
+          val (requests, newState, newDownloaderState) =
+            assignTasksToPeers(
+              NonEmptyList.fromListUnsafe(freePeers.toList),
+              currentState,
+              currentDownloaderState,
+              syncConfig.nodesPerRequest
             )
 
-            val (missing, state2) = sync.getMissingNodes(newState, downloaderCap)
+          requests.foreach(req => requestNodes(req))
+          processNodes(newState, currentStats, newDownloaderState, nodes).pipeTo(self)
+          context.become(
+            syncing(
```

Comment:
that syncing handler tracks quite a bit of state now, maybe it makes sense to extract it to some class?

Reply:
yep it will probably make code clearer